04-onion-routing.md

```diff
@@ -150,52 +151,50 @@ The overall structure of the packet is as follows:
 2. data:
    * [`1`:`version`]
    * [`33`:`public_key`]
-   * [`20*65`:`hops_data`]
+   * [`1300`:`hops_data`]
    * [`32`:`hmac`]
 
 For this specification (_version 0_), `version` has a constant value of `0x00`.
 
-The `hops_data` field is a structure that holds obfuscations of the next hop's
-address, transfer information, and its associated HMAC. It is 1300 bytes (`20x65`) long
-and has the following structure:
+The `hop_payloads` field is a structure that holds obfuscated routing information, and associated HMAC.
+It is 1300 bytes long and has the following structure:
 
-1. type: `hops_data`
+1. type: `hop_payloads`
 2. data:
-   * [`1`:`realm`]
-   * [`32`:`per_hop`]
+   * [`1-3`:`length`]
```

Comment:
We also cut off future upgrade paths by removing the realm here. Instead, we can keep the realm and use a fixed size 2-byte integer.

Reply:
Only if we want to upgrade out of TLV right?
If our stance is that TLV is the future we don't need a realm, the TLV `type` is the effective realm.
But if you think we need a way to move away from TLV in the future, in that case the realm would be needed.
I don't have a strong opinion on that, I think we still have the top-level onion version if we ever need to move away from TLV.

Reply:
I took a look at this again, and I think it's better to extract the common parts like the patch proposes, and then go into details further down. I'd like to treat the legacy version as a special case of the newer thing, rather than the other way around.

Reply:
> But if you think we need a way to move away from TLV in the future, in that case the realm would be needed.

I have no idea if we'll need to do so in the future, I don't think any of us do. Therefore, we should cautious to ensure we don't box ourselves into the current format without any way to transition away from it.

Reply:
That's reasonable :)
I believe that we have 2 ways of transitioning away, either with the `0x01` length or with the top-level onion version.
So we should be covered.